plugins/node/opentelemetry-instrumentation-router/src/instrumentation.ts

```diff
@@ -0,0 +1,211 @@
+          this._unwrap(Layer.prototype, 'handle_error');
+          return moduleExports;
+        }
+      )
+    );
+
+    return module;
+  }
+
+  // Define handle_request wrapper separately to ensure the signature has the correct length
+  private _requestHandlerPatcher(original: Router.Layer['handle_request']) {
+    const instrumentation = this;
+    return function wrapped_handle_request(
+      this: Router.Layer,
+      req: Router.RoutedRequest,
+      res: http.ServerResponse,
+      next: Router.NextFunction
+    ) {
+      if (utils.isInternal(this.handle) || this.handle.length > 3) {
+        return original.call(this, req, res, next);
+      }
+      const { context, wrappedNext } = instrumentation._setupSpan(
+        this,
+        req,
+        res,
+        next
+      );
+      return api.context.with(context, original, this, req, res, wrappedNext);
+    };
+  }
+
+  // Define handle_error wrapper separately to ensure the signature has the correct length
+  private _errorHandlerPatcher(original: Router.Layer['handle_error']) {
+    const instrumentation = this;
+    return function wrapped_handle_request(
+      this: Router.Layer,
+      error: Error,
+      req: Router.RoutedRequest,
+      res: http.ServerResponse,
+      next: Router.NextFunction
+    ) {
+      if (utils.isInternal(this.handle) || this.handle.length !== 4) {
+        return original.call(this, error, req, res, next);
+      }
+      const { context, wrappedNext } = instrumentation._setupSpan(
+        this,
+        req,
+        res,
+        next
+      );
+      return api.context.with(
+        context,
+        original,
+        this,
+        error,
+        req,
+        res,
+        wrappedNext
+      );
+    };
+  }
+
+  private _setupSpan(
+    layer: Router.Layer,
+    req: Router.RoutedRequest,
+    res: http.ServerResponse,
+    next: Router.NextFunction
+  ) {
+    const fnName = layer.handle.name || undefined;
+    const type = layer.method
+      ? types.LayerType.REQUEST_HANDLER
+      : types.LayerType.MIDDLEWARE;
+    const route = utils.getRoute(req);
+    const spanName =
+      type === types.LayerType.REQUEST_HANDLER
+        ? `request handler - ${route}`
+        : `middleware - ${fnName || '<anonymous>'}`;
+    const attributes = {
+      [types.CustomAttributeNames.NAME]: fnName,
+      [types.CustomAttributeNames.VERSION]: this._moduleVersion || 'n/a',
+      [types.CustomAttributeNames.TYPE]: type,
+      [SemanticAttributes.HTTP_ROUTE]: route,
+    };
+
+    const parent = api.context.active();
+    const parentSpan = api.getSpan(parent) as types.InstrumentationSpan;
+    const span = this.tracer.startSpan(
+      spanName,
+      {
+        attributes,
+      },
+      parent
+    ) as types.InstrumentationSpan;
+
+    utils.renameHttpSpan(parentSpan, layer.method, route);
+    // make sure spans are ended at least when response is finished
+    res.prependOnceListener('finish', () => span.end());
```

Comment:
I couldn't find the discussion in #416 but ending twice the span isn't technically a noop because its log a warning to the diag logger. Now that isn't enabled by default isn't a problem for all users but i suspect it will for people having it enabled (see https://github.com/open-telemetry/opentelemetry-js-contrib/issues/107 for exemple).

Reply:
I couldn't find the discussion in #416 but ending twice the span isn't technically a noop because its log a warning to the diag logger. 
Now that isn't enabled by default isn't a problem for all users but i suspect it will for people having it enabled (see https://github.com/open-telemetry/opentelemetry-js-contrib/issues/107 or https://github.com/open-telemetry/opentelemetry-js/pull/2126 for exemple).

Reply:
I'm going to add a check here, but I feel like it's such an widespread usecase, we should address it somehow better. **Any ideas?**

I cannot see a situation where this would even be something to be concerned about.